fix(core): deserialize malformed types in tx (#19446)

TheMrAI opened the pull request:
# Description of change

Based on [changes](https://github.com/MystenLabs/sui/commit/6d8ceed4d727a6c9ce9f5879b3cd1b2e8605affa).
(For original description see the ticket.)

Most APIs, command line tools simply do not allow the user to construct syntactically incorrect transaction, but there are APIs which do. Namely ones in which users can send in an already prepared transaction as a byte sequence.

```
Usage: iota client execute-signed-tx [OPTIONS] --tx-bytes <TX_BYTES>
```

or the GraphQL API:

```
[executeTransactionBlock](https://graphql.testnet.iota.cafe/#)(
txBytes: [String](https://graphql.testnet.iota.cafe/#)!
signatures: [[String](https://graphql.testnet.iota.cafe/#)!]!
): [ExecutionResult](https://graphql.testnet.iota.cafe/#)!
```

or the Json RPC API:

```
iota_executeTransactionBlock
```

In terms of this change the cases in which an invalid **module** or **function** identifier manages to sneak through the checks the system would have deemed it a valid transaction, but it would have failed during execution. Because of this
it would have cost gas, forcing it the invalid transaction to be stored on-chain. This would cause downstream interfaces
to fail while trying to access the invalid transaction.

For example calling this is a correct call for the clock timestamp:
`cargo run -p iota -- client call --package 0x2 --module clock --function timestamp_ms --args 0x6`
While this would be incorrect:
`cargo run -p iota -- client call --package 0x2 --module _ --function timestamp_ms --args 0x6`

Now if you tried to execute the second function you would notice that the call fails as the tool properly protects the user. But if you manage to construct a byte sequence containing the same function call with an underscore as a module name, you could sign it and send it in (through the raw byte interfaces), and get the above described issue.

This change aims to mitigate these issues. Now the **module** and **function** identifiers are checked for being syntactically valid identifiers, so after the new protocol version, no more transactions can be logged on-chain.
Furthermore the GraphQL API has been modified so that it won't crash while trying to parse such an invalid entry.

We have added an e2e test for checking the modified behavior. 

## Links to any relevant issues

fixes #7081 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Regenerate crates/iota-core/tests/staged/iota.yaml:
`cargo run -p iota-core --example generate-format -- print > crates/iota-core/tests/staged/iota.yaml`

Regenerate crates/iota-graphql-rpc/tests/snapshots/snapshot_tests__schema_sdl_export.snap
`cargo nextest run -p iota-graphql-rpc --features pg_integration --no-fail-fast --test-threads 1`

Regenerate/update iota-protocol-config snapshots:
`cargo nextest run -p iota-protocol-config`

Run iota-graphql-e2e-tests:
`cargo nextest run -p iota-graphql-e2e-tests --features pg_integration`

The new e2e tests:
`cargo nextest run -p iota-e2e-tests --test tx_bytes_validity_check`

### Infrastructure QA (only required for crates that are maintained by @iotaledger/infrastructure)

- [ ] Synchronization of the indexer from genesis for a network including migration objects.
- [ ] Restart of indexer synchronization locally without resetting the database.
- [ ] Restart of indexer synchronization on a production-like database.
- [ ] Deployment of services using Docker.
- [ ] Verification of API backward compatibility.

### Release Notes

<!--
Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates.
-->

- [x] Protocol: Transactions that include invalid types will not be
signed by validators from protocol 10 onwards.
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [x] GraphQL: `MoveType` can fail to provide a layout or abilities in
cases where it represents an invalid type.
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

amnn committed:
fix(core): deserialize malformed types in tx (#19446)

Make it possible to represent a transaction that includes a bad type (a
type that does not parse correctly). Such transactions will fail to
execute but may exist on-chain and so we need to be able to represent
them.

To avoid confusion in future, we will perform an early (signing) check
for type correctness, preventing more examples of this pattern from
appearing on-chain.

GraphQL has also been updated so that it can represent such a
transaction without emitting an error.

```
sui$ cargo simtest
sui$ cargo nextest run -p sui-graphql-rpc
sui$ cargo nextest run -p sui-graphql-e2e-tests
```

---

Check each box that your changes affect. If none of the boxes relate to
your changes, release notes aren't required.

For each box you select, include information after the relevant heading
that describes the impact of your changes that a user might notice and
any actions they must take to implement updates.

- [x] Protocol: Transactions that include unparseable types will not be
signed by validators from protocol 60 onwards.
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [x] GraphQL: `MoveType` can fail to provide a layout or abilities in
cases where it represents an unparseable type.
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

TheMrAI committed:
Add developer doc to  get_registry function

TheMrAI commented on crates/iota-core/src/generate_format.rs:
```diff
@@ -114,79 +140,105 @@ fn get_registry() -> Result<Registry> {
+    tracer.trace_value(&mut samples, &si).unwrap();
+
+    let event = iota_types::event::Event::random_for_testing();
+    tracer.trace_value(&mut samples, &event).unwrap();
```
Event sample added as it contains an Identifier.

miker83z replied:
Is this needed? I think the result would be the same if the Event was not explicitly put here.

TheMrAI replied:
Unfortunately it is, otherwise the tracer will fail as it will not find a valid sample for the Identifier inside the Event.
It will try to serialize an empty string, upon which it will fail as that is forbidden within the Identifier.

miker83z replied:
Tested locally after the rebase, I still don't get why on Sui a sample for the Identifier is not needed here.

miker83z replied:
Could you point me to the part of the Event that is different in respect to upstream?

TheMrAI replied:
You are right. I confused it with something else. Investigating.

TheMrAI replied:
So for the time being I have checked the serde-reflection versions.
If I downgrade to serde-reflection 0.3.6 (same if I exactly copy the version pattern from sui and use 0.3.6 globally but for external crates 0.3.2) then we get error:

```
thread 'main' panicked at crates/iota-core/src/generate_format.rs:208:46:
called `Result::unwrap()` on an `Err` value: Custom("Failed to deserialize value: \"invalid value: integer `11`, expected variant index 0 <= i < 11\"") 
```
And it would point to the `tracer.trace_type::<TypeInput>(&samples).unwrap();` which is rather odd. Has 11 variants yes, why is it trying to serialize with the 11th?

If I upgrade to serde-reflection 0.5 then we get different error: 
```
thread 'main' panicked at crates/iota-core/src/generate_format.rs:318:35:
called `Result::unwrap()` on an `Err` value: MissingVariants(["CheckpointCommitment", "ConsensusDeterminedVersionAssignments", "GenesisObject", "TransactionData"])
```

But Sui is on version 0.3.6 and they don’t need Event sample and even if I update it to
0.4 it still passes. 
If I update their version to 0.5 they will error with:
```
thread 'main' panicked at crates/sui-core/src/generate_format.rs:248:35:
called `Result::unwrap()` on an `Err` value: MissingVariants(["CheckpointCommitment", "ConsensusDeterminedVersionAssignments", "GenesisObject", "TransactionData"])
```

As of now it is really unclear why this discrepancy occurs, nor is it trivially visible which type in the chain may cause an issue. My best guess is that perhaps serde-reflection isn't entirely stable and some subtle type differences may cause the problems.

valeriyr replied:
@TheMrAI @miker83z Looks like the problem is in the `GenesisTransaction` type, it doesn't contain events in Sui.
So, I think, the correct way to fix this issue should be to move the event tracing above this line:
```
tracer.trace_type::<TransactionKind>(&samples).unwrap();
```

TheMrAI replied:
Agreed @valeriyr, moved the line.

miker83z replied:
@TheMrAI we can close #7384, no?

TheMrAI commented on crates/iota-graphql-rpc/src/types/query.rs:
```diff
@@ -560,6 +559,15 @@ impl Query {
             }))
     }
 
+    // TODO <- patch this after [aebab3c](https://github.com/MystenLabs/sui/commit/aebab3cf3218a8893417209fecc834c23619e780#diff-dc915efeed8b11d1f192223812017db449f824f2df0043dd95e207e75b0ba70e)
```
Do we want such a comment to be included?
I have put in as a marker, but I don't like it.

miker83z replied:
I guess it is fine for now!

kodemartin replied:
I would rather suggest to create a tracking issue and remove the comment.

TheMrAI commented on crates/iota-package-resolver/src/lib.rs:
```diff
@@ -1727,6 +1732,38 @@ fn ident(s: &str) -> Result<Identifier> {
     Identifier::new(s).map_err(|_| Error::NotAnIdentifier(s.to_string()))
 }
 
+pub fn as_type_tag(type_input: &TypeInput) -> Result<TypeTag> {
```
!!! This function is the same function as as_type_tag and into_type_tag from crates/iota-types/src/type_input.rs.
This triplication is very unnecessary and error prone. What would you prefer?
We keep one `as_type_tag(&self)` implemented in type_inputs.rs or the more Rustecean TryFrom (or TryInto if we don't have visibility).

I think it is prudent to stay with as_type_tag(&self) to keep future pulls a bit simpler. WDYT?

miker83z replied:
I vote for moving the logic into 1 new function and keeping the 3 function signatures as these are (pls, first check if maybe ths was done in a more recent upstream commit)

TheMrAI replied:
Only managed to eliminate one of the triplets. This one would be too costly to fix.
For details see the Commits.

TheMrAI commented:
Please take care while reviewing this PR. I can't possible guarantee that no invariants got broken.

jkrvivian commented on crates/iota-types/src/type_input.rs:
```diff
@@ -0,0 +1,378 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2024 IOTA Stiftung
```
```suggestion
// Modifications Copyright (c) 2025 IOTA Stiftung
```

TheMrAI committed:
Update licence date to 2025

TheMrAI committed:
Remove duplicated into_type_tag implementation

Unfortunatelly the as_type_tag implementation from
crates/iota-package-resolver/src/lib.rs cannot be removed without moderately complicated refactoring.
The issue is that iota-package-resolver uses a custom error type while the iota-types implementation
is relying on a generic anyhow::Error.
We can't convert anyhow::Error to Error either as there is no way to access the raw data.

Similarly we could write a third mapping function which could return just the incoming faulty
Identifier string (as of now only Identifier parsing can fail) and reconstruct the appropriate error
at the right place. But while it may work now, it would be a nightmare to maintain if and when different
error types can be produced.
For now we keep the duplication as there is no future proof, cheap solution.

Thoralf-M commented on crates/iota-types/src/iota_sdk2_conversions.rs:
```diff
@@ -1766,21 +1778,19 @@ impl TryFrom<Command> for crate::transaction::Command {
+                crate::Identifier::new(move_call.module.as_str())
+                    .expect("invalid move call module identifier"),
+                crate::Identifier::new(move_call.function.as_str())
+                    .expect("invalid move call function identifier"),
```
Why expect here and not return an error?

TheMrAI replied:
That behavior wasn't changed as per the feature pull. If at all possible we shouldn't change code when pulling changes or did you mean something else?

Thoralf-M replied:
Ah it's the Command of the other SDK, then expect should be fine

miker83z commented on sdk/typescript/src/graphql/generated/2025.2/schema.graphql:
Please, remove these changes and create an issue for the tooling team to update this schema.

TheMrAI replied:
Of course, but could you provide more context why?
I am not even sure why this change is here.

TheMrAI replied:
Never mind, got the context.

miker83z commented on crates/iota-graphql-rpc/schema.graphql:
```diff
@@ -343,7 +343,6 @@ String representation of an arbitrary width, possibly signed integer.
 """
 scalar BigInt
 
-
```
Could you check if, when producing the schema locally, it contains these blank spaces or not?

TheMrAI replied:
It does not, I had to regenerate the schema and those spaces got deleted.

TheMrAI commented on crates/iota-core/src/generate_format.rs:
```diff
@@ -132,79 +150,105 @@ fn get_registry() -> Result<Registry> {
+    };
+    tracer.trace_value(&mut samples, &si).unwrap();
+
+    let event = iota_types::event::Event::random_for_testing();
```
This Event sample shouldn't be necessary, but we must put it in otherwise serde-reflection will fail for us.
There is already an Event at the bottom of this file, we could move that up here or just leave a comment with this that it is here to workaround serde-reflection inconsistencies.

Now supplying another sample will not break anything, so the presence of this line shouldn't affect the outputs.

TheMrAI commented:
Please let me know what other tests should be executed.

kodemartin commented on crates/iota-graphql-rpc/src/types/move_type.rs:
```diff
@@ -143,87 +143,98 @@ impl MoveType {
-    async fn abilities(&self, ctx: &Context<'_>) -> Result<Vec<MoveAbility>> {
+    /// The abilities this concrete type has. Returns no abilities if the type
+    /// is invalid.
+    async fn abilities(&self, ctx: &Context<'_>) -> Result<Option<Vec<MoveAbility>>> {
```
You can return an empty `Vec` if the type is invalid. This way we preserve the existing guarantees of the API.

kodemartin commented on crates/iota-graphql-rpc/src/types/move_type.rs:
```diff
@@ -143,87 +143,98 @@ impl MoveType {
     /// Structured representation of the "shape" of values that match this type.
-    async fn layout(&self, ctx: &Context<'_>) -> Result<MoveTypeLayout> {
+    /// May return no layout if the type is invalid.
+    async fn layout(&self, ctx: &Context<'_>) -> Result<Option<MoveTypeLayout>> {
```
Instead of making this a nullable value for the API, I would suggest that you add a `MoveTypeLayout::InvalidType` variant and return that instead of `None`.

kodemartin reviewed (changes requested):
Left a couple of suggestion requiring minor revisions in `iota-graphql-rpc`. Other than that it looks good for the infra crates.

TheMrAI committed:
Remove change marker from query.rs

It makes more sense to keep such comments out of the source as it
is unclear if we ever even gonna pull the change which it modifies.

TheMrAI committed:
Restore iota-graphql-rpc schema

With this graphql schema remains the same, while also gracefully
handling the malformed types.

TheMrAI committed:
Add explanation comment for generate_format hack

We do not know why the inconsistency appears. The most likely
scenario seems to be that serde-reflection is not entirely stable,
or that we have changed something subtly with our types which causes
cascading issues due to the nature of serde-reflection's algorithm.

kodemartin commented on crates/iota-graphql-rpc/src/types/move_type.rs:
```diff
@@ -143,87 +144,98 @@ impl MoveType {
         resolver: &PackageResolver,
-    ) -> Result<AbilitySet, Error> {
-        resolver.abilities(self.native.clone()).await.map_err(|e| {
+    ) -> Result<Option<AbilitySet>, Error> {
```
Nit: If you maintain the signature here as well you might not need to refactor the calling function above at all.

TheMrAI replied:
I see your point, but am rather torn.

For one I try to stay as close to the feature pull as possible, even if I don't like it.
If I keep the original interface, then I can't really signal this new special case, because here it could only be an Error (in which case I would have to introduce a new type or match on its context above so that the empty vector could be returned).
Since, these aren't really that `pub` public functions, I could refactor abilities_impl, abilities function and the rest of the call sites to make it cleaner, but I feel that isn't in the scope of this PR.

The similar goes to the above layout you mentioned.

What do you think? I am not against fixing it, but probably would be best done in a follow up PR. This one is already a bit too complicated.

kodemartin commented on crates/iota-graphql-rpc/src/types/move_type.rs:
```diff
@@ -143,87 +144,98 @@ impl MoveType {
-                    self.native.to_canonical_display(/* with_prefix */ true),
-                ))
-            })
+    ) -> Result<Option<A::MoveTypeLayout>, Error> {
```
Nit: If you maintain the signature here as well you might not need to refactor the calling function above at all.

kodemartin reviewed (approved):
looks good @TheMrAI ! Left a couple of nitpicks that you may or may not consider further. Thank you!

TheMrAI committed:
Add missing comment to protocol version history comment

Forgot to add this and after noticing others doing it, added it.

miker83z reviewed (approved):
LGTM!

valeriyr commented on crates/iota-core/src/generate_format.rs:
```diff
@@ -56,6 +57,23 @@ use serde_reflection::{Registry, Result, Samples, Tracer, TracerConfig};
+///
+/// For example **TypeA** implements a custom serializer, hence necessitating
+/// the use of [Tracer::trace_value], then every type that contains **TypeA**
+/// will require a sample to be provided.
```
Should we change the line:
```
// tracer.trace_value(&mut samples, ...)?;
```
to
```
// tracer.trace_value(&mut samples, ...).unwrap();
```
in the comment below?

TheMrAI replied:
We can, but it will still be misleading, because the with the changes introduced in: https://github.com/iotaledger/iota/commit/ef3cd55a6427cdbf0c111336f4882cb94727f6d6
we already broke the separation that there is a trace values section followed by a trace types section.
This is how the code is in sui, and I have attempted to resolve the order for us, but it would introduce other tracing failures which didn't seem worth fixing.

Rather difficult to fix any problem with the current constraints.

TheMrAI replied:
Not to mention that unwrapping is a downgrade in all honesty. When I was tracking down the issue it turned out that you can ask serde to provide some explanation for the failure which was admittedly a bit generic, but helpful non the less.
However, with this panicing unwrap, such niceties are lost.

valeriyr reviewed (approved):
The only point is that the extra event tracing can be removed. Otherwise, everything looks good to me.

TheMrAI committed:
Move Event sample up in generate_format.rs

This fixes the issue a bit more cleanly, but it is still a bit confusing,
so the comment tries to indicate the why.

TheMrAI committed:
Update comment about sample unwrapping

TheMrAI committed:
feat(iota-e2e-tests): add e2e test for invalid Identifier check (#7488)

Adding an e2e test checking whether or not a transaction constructed
containing invalid module or function Identifiers can be sent in.

In protocol version 9 these can still be accepted, but from version 10
they should be rejected outright.

### Tests exucted

```cargo nextest run -p iota-e2e-tests --test tx_bytes_validity_check```

marc2332 committed:
refactor(ts-sdk): Sync ts-sdk with 'deserialize malformed types in tx' (#7439)

Closes https://github.com/iotaledger/iota/issues/7434

[run-ci]
